### PR TITLE
Add the SDK config generator for OpenNSA

### PR DESCRIPTION
**Pull request by pudelkoM**
SDK configuratibility is achieved by moving the `BcmChassisManager::WriteBcmConfigFile` member function to a SDK specific implementation file.

### CODE REVIEW

File: bazel/rules/library_rule.bzl [hunk 1]
```diff
@@ -25,9 +25,11 @@ def stratum_cc_library(name, deps = None, srcs = None, data = None,
 
   alwayslink = 0
   if srcs:
-    if [s for s in srcs if not s.endswith(".h")]:
+    if type(srcs) == "select":
```
Comment (pudelkoM) [inline bazel/rules/library_rule.bzl]: This is needed for `select` sources. I looked through all the targets we build on CI, and all of them had `      alwayslink = 1` applied. So I think it's save to do in this case.

File: stratum/hal/lib/bcm/bcm_chassis_manager_config_generator_sdk6.cc [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    }
+    // XE ports polarity flip handling for RX and TX.
+    if (bcm_port.tx_polarity_flip() > 0) {
+      buffer << "phy_xaui_tx_polarity_flip_xe" << bcm_port.diag_port() << "."
+             << bcm_port.unit() << "=0x" << std::hex << std::uppercase
+             << bcm_port.tx_polarity_flip() << std::dec << std::nouppercase
+             << std::endl;
+    }
+    if (bcm_port.rx_polarity_flip() > 0) {
+      buffer << "phy_xaui_rx_polarity_flip_xe" << bcm_port.diag_port() << "."
```
Comment (pudelkoM) [inline stratum/hal/lib/bcm/bcm_chassis_manager_config_generator_sdk6.cc]: Let's not use the `phy_xaui_rx_polarity_flip_xe` format, but logical ports instead.

File: stratum/hal/lib/bcm/BUILD [hunk 3]
```diff
@@ -199,7 +199,11 @@ stratum_cc_library(
         "//stratum/public/lib:error",
         "//stratum/glue/gtl:map_util",
         "//stratum/glue/gtl:stl_util",
-    ],
+    ] + select({
+        ":bcm_sdk6": [],
+        ":bcm_sdklt": ["@com_github_jbeder_yaml_cpp//:yaml_cpp"],
+        "//conditions:default": ["@com_github_jbeder_yaml_cpp//:yaml_cpp"],
+    }),
```
Comment (pudelkoM) [inline stratum/hal/lib/bcm/BUILD]: I see the drawbacks of my approach now. Maybe factoring the generator out into it's own target is not so bad after all...

File: stratum/hal/lib/bcm/bcm_chassis_manager_config_generator_sdklt.cc [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    case 42000000000: return "PC_PORT_OPMODE_42G";
+    case 50000000000: return "PC_PORT_OPMODE_50G";
+    case 100000000000: return "PC_PORT_OPMODE_100G";
+    case 120000000000: return "PC_PORT_OPMODE_120G";
+    case 127000000000: return "PC_PORT_OPMODE_127G";
+    case 200000000000: return "PC_PORT_OPMODE_200G";
+    case 400000000000: return "PC_PORT_OPMODE_400G";
+    default: return "PC_PORT_OPMODE_ANY";
+  }
+}
```
Comment (pudelkoM) [inline stratum/hal/lib/bcm/bcm_chassis_manager_config_generator_sdklt.cc]: Can this be moved into an anonymous namespace?

File: stratum/hal/lib/bcm/bcm_chassis_manager_config_generator_sdklt.cc [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    const BcmChassisMap& base_bcm_chassis_map,
+    const BcmChassisMap& target_bcm_chassis_map) const {
+  std::stringstream buffer;
+
+  absl::flat_hash_map<int, SerdesLaneSetting*> serdes_lane_settings;
+  for (const auto& bcm_port : target_bcm_chassis_map.bcm_ports()) {
+    int serdes_core_id = bcm_port.serdes_core();
+
+    if (!serdes_lane_settings.count(serdes_core_id)) {
+      serdes_lane_settings.emplace(serdes_core_id, new SerdesLaneSetting());
```
Comment (pudelkoM) [inline stratum/hal/lib/bcm/bcm_chassis_manager_config_generator_sdklt.cc]: nice memory leak

Comment (pudelkoM) [inline stratum/hal/lib/bcm/bcm_chassis_manager_config_generator_sdklt.cc]: nice memory leak

File: stratum/hal/lib/bcm/bcm_chassis_manager_config_generator_sdklt.cc [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+    }
+
+    SerdesLaneSetting* serdes_lane_setting =
+        serdes_lane_settings[serdes_core_id];
+
+    if (!serdes_lane_setting) {
+      // This should not happened
+      RETURN_ERROR() << "Serdes core id " << serdes_core_id << " not found.";
+    }
```
Comment (pudelkoM) [inline stratum/hal/lib/bcm/bcm_chassis_manager_config_generator_sdklt.cc]: `CHECK_RETURN_IF_FALSE`

File: stratum/hal/lib/bcm/bcm_chassis_manager_config_generator_sdklt.cc [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+    SerdesLaneSetting* serdes_lane_setting = serdes_lane_settings[pc_pm_id];
+
+    // Key is a map (PC_PM_ID: xx, CORE_INDEX: unit)
+    pc_pm_core << YAML::Key << YAML::BeginMap << YAML::Key << "PC_PM_ID"
+               << YAML::Value << pc_pm_id << YAML::Key << "CORE_INDEX"
+               << YAML::Value << bcm_port.unit() << YAML::EndMap;
+
+    pc_pm_core << YAML::Value << YAML::BeginMap;
+    pc_pm_core << YAML::Key << "RX_LANE_MAP";
+    pc_pm_core << YAML::Value << serdes_lane_setting->rx_lane_map;
```
Comment (pudelkoM) [inline stratum/hal/lib/bcm/bcm_chassis_manager_config_generator_sdklt.cc]:
```suggestion
    pc_pm_core << YAML::Value << YAML::BeginMap
        << YAML::Key << "RX_LANE_MAP"
        << YAML::Value << serdes_lane_setting->rx_lane_map;
```

**Comment by pudelkoM** [comment]
Needs to be updated after PR #155

**Comment by Yi-Tseng** [comment]
I guess we plan to merge this after the OpenNSA PR, right?

**Comment by pudelkoM** [comment]
I think so. In the OpenNSA PR we decide how to integrate both SDKs, then we apply this here.

**Comment by bocon13** [comment]
We can merge this into the opennsa branch if you'd like

**Comment by pudelkoM** [comment]
Looks good, just one thing: We could remove the `_sdk` suffix from the generator filename, since they reside in different directories now.

**Pull request merged**